orm/src/main/java/org/hibernate/search/v6poc/entity/orm/model/impl/HibernateOrmPathFilterFactory.java

```diff
@@ -0,0 +1,290 @@
+ *         and is not an association
+ *         (i.e. {@link SimpleValue} but not its subtypes),
+ *         then the path will be represented as the "property path" of that prefix.
+ *
+ *         Rationale: such values are considered as atomic when persisting,
+ *         i.e. they will be reported by ORM as dirty in a {@link PostUpdateEvent}
+ *         whenever any of their parts changed.
+ *     </li>
+ *     <li>
+ *         Otherwise, if a prefix of the path can be resolved to a multi-valued {@link Value}
+ *         (i.e. a {@link org.hibernate.mapping.Collection}, which also encompasses maps)
+ *         containing only embeddeds (i.e. {@link Component})
+ *         or "atomic" values (i.e. {@link SimpleValue} but not its subtypes),
+ *         then the path will be represented as either the "property path" of that prefix
+ *         and alternatively as the collection role.
+ *
+ *         Rationale: such values are considered as atomic when persisting,
+ *         i.e. they will be reported by ORM as dirty in an {@link org.hibernate.event.spi.AbstractCollectionEvent}
+ *         whenever any of their parts changed.
+ *     </li>
+ *     <li>
+ *         Otherwise, no string representation can be assigned and an exception will be thrown.
+ *         This includes in particular all cases when the path points to a {@link javax.persistence.Transient} property,
+ *         or when a custom or {@link java.util.Optional} value extractor
+ *         is used before we can detect a prefix matching the conditions described above.
+ *     </li>
+ * </ul>
+ */
+public class HibernateOrmPathFilterFactory implements PojoPathFilterFactory<Set<String>> {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final PersistentClass persistentClass;
+
+	public HibernateOrmPathFilterFactory(PersistentClass persistentClass) {
+		this.persistentClass = persistentClass;
+	}
+
+	@Override
+	public PojoPathFilter<Set<String>> create(Set<PojoModelPathValueNode> paths) {
+		// Use a LinkedHashSet for deterministic iteration
+		Set<String> pathsAsStrings = new LinkedHashSet<>();
+		for ( PojoModelPathValueNode path : paths ) {
+			addDirtyPathStringRepresentations( pathsAsStrings, path );
+		}
+		return new StringSetPojoPathFilter( pathsAsStrings );
+	}
+
+	private void addDirtyPathStringRepresentations(Set<String> pathsAsStrings, PojoModelPathValueNode path) {
+		Optional<Value> valueOptional = addDirtyPathStringRepresentationsRecursively( pathsAsStrings, path, true );
+		if ( valueOptional.isPresent() ) {
+			Value value = valueOptional.get();
+			/*
+			 * We were able to resolve the path, but didn't find any Value that could possibly
+			 * be reported as dirty by Hibernate ORM.
+			 */
+			throw log.unreportedPathForDirtyChecking( persistentClass.getMappedClass(), path, value );
+		}
+		// Else everything is good, the string representation was successfully added to the set.
```

Comment:
I find the logic weird. I wouldn't have returned a path and would have throw the exception in the method below.

Reply:
We might have two paths in some cases, so I can't return a path.

I could throw the exception in the method below, but I suspect it will make the code even more complex: I would have to perform this check (`if (isWholePath)`) everywhere we currently do `return Optional.of( value )`...